### ISSUE TITLE
[processor/k8sattributes] Add support for getting metadata from Kubelet 

### ISSUE THREAD

**Issue by povilasv** [open-telemetry/opentelemetry-collector-contrib#14475]
### Is your feature request related to a problem? Please describe.

With large Kubernetes clusters k8s attribute processor might send quite a lot of requests to the Kubernetes API Server, which can cause problems. Example reported by Fluent Bit -> https://docs.fluentbit.io/manual/pipeline/filters/kubernetes#optional-feature-using-kubelet-to-get-metadata

### Describe the solution you'd like

Similarly to Fluent Bit we can add a "Use Kubelet" feature, which would call Kubelet /pods API to get Pod metadata :)

Reference: https://docs.fluentbit.io/manual/pipeline/filters/kubernetes#optional-feature-using-kubelet-to-get-metadata

**Comment by fatsheep9146** [comment]
I will try to solve this issue.

@TylerHelmuth @dmitryax

**Comment by cdanis** [comment]
At the Wikimedia Foundation, we recently suffered a Kubernetes control plane capacity crunch that was partially triggered by enabling k8sattributesprocessor on our otelcol daemonset.

There was no impact to our users, but deploys of new code were blocked for some time, and we spent several engineer-days debugging and fixing.

This would have been very likely mitigated by having the kubelet /pods api usable by processor/k8sattributes.

A full writeup of the incident is available at https://wikitech.wikimedia.org/wiki/Incidents/2024-05-28_wikikube-api

**Comment by TylerHelmuth** [comment]
@cdanis do you have any interest in working on this feature?

**Comment by bacherfl** [comment]
@TylerHelmuth if this issue is still available, I would like to look into it

**Comment by TylerHelmuth** [comment]
Yes

**Comment by bacherfl** [comment]
I have been looking into this now, and came up with a draft PR to illustrate a possible implementation. Please note that at this stage this is only a very rough draft with a couple of hard coded parameters/urls, but it should serve as a base for a discussion on how to best proceed with this issue.

Below is a list of the most important findings I have made so far:

- The kubelet API is not officially documented and potentially prone to change. The best source of documentation in this case is the [code of the kubelet](https://github.com/kubernetes/kubernetes/blob/bd239d42e463bff7694c30c994abd54e4db78700/pkg/kubelet/server/server.go)
- The Kubernetes [client-go](https://github.com/kubernetes/client-go) library does not provide support for directly interacting with the kubelet API - There is the possibility to reach the kubelet's endpoints via the client's `WrapTransport` option where the request URL can be set to the `/pod` endpoint of the kubelet, so the authentication configuration of the kubernetes client can be reused, but this is a bit of a hack, and comes with several limitations, as outlined in the following points
- The kubelet API only provides a really simple endpoint for retrieving all pods (or only the running pods) on the node. The structure of the response is the same as for the pods list endpoint of the kubernetes API, but the kubelet endpoint does not provide any filtering capabilities. So the Label/Namespace/Attribute filters will all have to be applied in the collector after retrieving the list of Pods
- No support for using Informers. The current implementation relies on the use of informers that call the `Watch` function of the client to receive an event every time one of the observed resources is added/updated/deleted (e.g. when a pod is added or deleted). This is not possible with the kubelet API, so we'd have to rely on polling the `/pod` endpoint regularly, and synchronise the received list of pods with the latest known status of the collector. This comes with quite an overhead if we want to maintain the `Added|Updated|DeletedPods` counters which are exposed as metrics by the collector as we have to compare the retrieved pods against the list of pods the collector currently has in memory to know if any pods have been deleted or added (see the diff of the PR).

To summarise, I think this could potentially be implemented, but as a experimental feature with a number of drawbacks which need to be well documented. @TylerHelmuth and others FYI - when you have some time feel free to have a look at the draft PR and the findings listed above - appreciate any input on how to (or if) best proceed with that issue

**Comment by ChrsMark** [comment]
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/44787 seems to be a duplicate of this.

**Comment by iblancasa** [comment]
> [#49427](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/49427) seems to be a duplicate of this.

That one is a PR proposal for solving this ticket.
